builder/dockerfile/dispatchers.go

```diff
@@ -208,65 +141,70 @@ func (b *Builder) getImageMount(fromFlag *Flag) (*imageMount, error) {
 
 // FROM imagename[:tag | @digest] [AS build-stage-name]
 //
-func from(req dispatchRequest) error {
-	stageName, err := parseBuildStageName(req.args)
-	if err != nil {
-		return err
-	}
-
-	if err := req.flags.Parse(); err != nil {
-		return err
-	}
-
-	req.builder.imageProber.Reset()
-	image, err := req.builder.getFromImage(req.shlex, req.args[0])
+func (d *stageBuilder) dispatchFrom(cmd *instructions.FromCommand) error {
+	d.builder.imageProber.Reset()
+	image, err := d.builder.getFromImage(d.shlex, cmd.BaseName)
 	if err != nil {
 		return err
 	}
-	if err := req.builder.buildStages.add(stageName, image); err != nil {
+	if err := d.builder.buildStages.add(cmd.StageName, image); err != nil {
 		return err
 	}
-	req.state.beginStage(stageName, image)
-	req.builder.buildArgs.ResetAllowed()
-	if image.ImageID() == "" {
-		// Typically this means they used "FROM scratch"
-		return nil
-	}
-
-	return processOnBuild(req)
-}
-
-func parseBuildStageName(args []string) (string, error) {
-	stageName := ""
-	switch {
-	case len(args) == 3 && strings.EqualFold(args[1], "as"):
-		stageName = strings.ToLower(args[2])
-		if ok, _ := regexp.MatchString("^[a-z][a-z0-9-_\\.]*$", stageName); !ok {
-			return "", errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", stageName)
+	state := d.state
+	state.beginStage(cmd.StageName, image)
+	d.builder.buildArgs.ResetAllowed()
+	if state.runConfig.OnBuild != nil && len(state.runConfig.OnBuild) > 0 {
```

Comment:
I think we should keep the `OnBuild` handling as a separate function. It makes the individual steps of `dispatchFrom()` a lot more clear